### PR TITLE
Add e2e testing

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

This PR introduces e2e testing using uptest.

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR, as appropriate.

### How has this code been tested

```shell
export UPTEST_AWS_CREDS=<REDACTED>

make e2e
```

Once this PR got merged, upcoming PR could be tested by adding a comment with `/test-e2e`

### CODE REVIEW

**Comment by turkenh** [comment]
> Left some comments on CI..
> 
> I'm hitting locally
> 
> ```
>    logger.go:42: 14:15:34 | case/0-apply | error: An error occurred while waiting for the condition to be satisfied: an error on the server ("unable to decode an event from the watch stream: http2: client connection lost") has prevented the request from succeedingUnable to connect to the server: net/http: TLS handshake timeout
>     logger.go:42: 14:15:34 | case/0-apply | Unable to connect to the server: net/http: TLS handshake timeout
> 
>     case.go:361: failed in step 0-apply
>     case.go:363: exit status 1
>     case.go:363: Get "https://127.0.0.1:54000/api?timeout=32s": net/http: TLS handshake timeout
>     logger.go:42: 14:15:44 | case | Failed to collect events for case in ns kuttl-test-diverse-swan: func passed to retry panicked.  expected if testsuite is shutting down
> ```
> 
> Which seems to be local kind performance related... Have you seen it before?

I didn't see it but I would suspect the same. I am also wondering how the CI with Github Action behave regarding the performance which I would see after merge.

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+      - name: Checkout PR
+        id: checkout-pr
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh pr checkout ${{ github.event.issue.number }}
+          git submodule update --init --recursive
+          OUTPUT=$(git log -1 --format='%H')
+          echo "::set-output name=commit-sha::$OUTPUT"
```
[the remark this answers is not in the document]
Comment (ytsarev) [reply .github/workflows/e2e.yaml:61]: tracked in https://github.com/upbound/uptest/issues/31

File: .github/workflows/e2e.yaml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh api \
+            --method POST \
+            -H "Accept: application/vnd.github+json" \
+            /repos/${{ github.repository }}/statuses/${{ steps.checkout-pr.outputs.commit-sha }} \
+            -f state='success' \
+            -f target_url='https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}' \
+            -f description='Passed' \
+            -f context="E2E"
```
[the remark this answers is not in the document]
Comment (ytsarev) [reply .github/workflows/e2e.yaml:82]: tracked in https://github.com/upbound/uptest/issues/31

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+      - name: Checkout PR
+        id: checkout-pr
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh pr checkout ${{ github.event.issue.number }}
+          git submodule update --init --recursive
+          OUTPUT=$(git log -1 --format='%H')
+          echo "::set-output name=commit-sha::$OUTPUT"
```
[the remark this answers is not in the document]
Comment (turkenh) [reply .github/workflows/e2e.yaml:28]: Makes sense to fix as part of https://github.com/upbound/uptest/issues/31

**Comment by turkenh** [comment]
> Left some comments on CI..
> 
> I'm hitting locally
> 
> ```
>    logger.go:42: 14:15:34 | case/0-apply | error: An error occurred while waiting for the condition to be satisfied: an error on the server ("unable to decode an event from the watch stream: http2: client connection lost") has prevented the request from succeedingUnable to connect to the server: net/http: TLS handshake timeout
>     logger.go:42: 14:15:34 | case/0-apply | Unable to connect to the server: net/http: TLS handshake timeout
> 
>     case.go:361: failed in step 0-apply
>     case.go:363: exit status 1
>     case.go:363: Get "https://127.0.0.1:54000/api?timeout=32s": net/http: TLS handshake timeout
>     logger.go:42: 14:15:44 | case | Failed to collect events for case in ns kuttl-test-diverse-swan: func passed to retry panicked.  expected if testsuite is shutting down
> ```
> 
> Which seems to be local kind performance related... Have you seen it before?

@ytsarev added some more checks to `test/setup.sh`. Would you mind giving this another try? 
I would like to merge and test everything e2e with CI.

**Comment by ytsarev** [comment]
@turkenh I got that locally, looks better but times out

```
    logger.go:42: 22:27:02 | case/0-apply | Running setup.sh
    logger.go:42: 22:27:02 | case/0-apply | Waiting until configuration package is healthy/installed...
    logger.go:42: 22:32:03 | case/0-apply | error: timed out waiting for the condition on configurations/platform-ref-aws
I1019 22:32:04.237280   20454 request.go:655] Throttling request took 1.005720666s, request: GET:https://0.0.0.0:64302/apis/sql.gcp.upbound.io/v1beta1?timeout=32s
    logger.go:42: 22:32:10 | case/0-apply | Cluster:default/platform-ref-aws created
    case.go:361: failed in step 0-apply
    case.go:363: exit status 1
Warning: events.k8s.io/v1beta1 Event is deprecated in v1.22+, unavailable in v1.25+
    logger.go:42: 22:32:11 | case | case events from ns kuttl-test-living-wildcat:
    logger.go:42: 22:32:11 | case | Deleting namespace: kuttl-test-living-wildcat
=== CONT  kuttl
    harness.go:402: run tests finished
    harness.go:511: cleaning up
    harness.go:553: skipping cluster tear down
    harness.go:554: to connect to the cluster, run: export KUBECONFIG="/Users/xnull/upbound/platform-ref-aws/kubeconfig"
--- FAIL: kuttl (331.52s)
    --- FAIL: kuttl/harness (0.00s)
        --- FAIL: kuttl/harness/case (308.36s)
FAIL
uptest-v0.1.1: error: cannot run e2e tests successfully: cannot execute tests: kuttl failed: exit status 1
```

Let's see how it behaves in GHA

**Review by ytsarev — APPROVED** [review]

**Pull request merged**